cmd/kubeless/function/top.go

```diff
@@ -0,0 +1,270 @@
+/*
+Copyright (c) 2016-2017 Bitnami
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package function
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/gosuri/uitable"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	kubelessApi "github.com/kubeless/kubeless/pkg/apis/kubeless/v1beta1"
+	"github.com/kubeless/kubeless/pkg/client/clientset/versioned"
+	"github.com/kubeless/kubeless/pkg/utils"
+
+	"github.com/prometheus/common/expfmt"
+)
+
+var topCmd = &cobra.Command{
+	Use:     "top",
+	Aliases: []string{"stats"},
+	Short:   "display function metrics",
+	Long:    `display function metrics`,
+	Run: func(cmd *cobra.Command, args []string) {
+		functionName, err := cmd.Flags().GetString("function")
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		ns, err := cmd.Flags().GetString("namespace")
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		if ns == "" {
+			ns = utils.GetDefaultNamespace()
+		}
+		output, err := cmd.Flags().GetString("out")
+		if err != nil {
+			logrus.Fatal(err.Error())
+		}
+
+		apiV1Client := utils.GetClientOutOfCluster()
+		kubelessClient, err := utils.GetKubelessClientOutCluster()
+		handler := &PrometheusMetricsHandler{}
+
+		err = doTop(cmd.OutOrStdout(), kubelessClient, apiV1Client, handler, ns, functionName, output)
+		if err != nil {
+			logrus.Fatal(err.Error())
+		}
+	},
+}
+
+// Metric contains metrics for a functions
+type Metric struct {
+	FunctionName         string  `json:"function,omitempty"`
+	Namespace            string  `json:"namespace,omitempty"`
+	Method               string  `json:"method,omitempty"`
+	TotalCalls           float64 `json:"total_calls,omitempty"`
+	TotalFailures        float64 `json:"total_failures,omitempty"`
+	TotalDurationSeconds float64 `json:"total_duration_seconds,omitempty"`
+	AvgDurationSeconds   float64 `json:"avg_duration_seconds,omitempty"`
+}
+
+func init() {
+	topCmd.Flags().StringP("namespace", "n", "", "Specify namespace for the function")
+	topCmd.Flags().StringP("function", "f", "", "Specify the function")
+	topCmd.Flags().StringP("out", "o", "", "Output format. One of: json|yaml")
+}
+
+// MetricsRetriever is an interface for retreiving metrics from an endpoint
```

Comment:
I would move all the code not related to the command line (from here to the bottom) to a different file to be easier to find and share